### PR TITLE
Disable Constraint constructor

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107918

### CODE REVIEW

**Review by zhxchen17 — APPROVED** [review]

File: torch/export/__init__.py [hunk 1]
```diff
@@ -597,11 +595,34 @@ class Constraint(_ConstraintTarget):
...
+        raise NotImplementedError(
+            "Can't directly construct Constraint object. Please use torch.export.dynamic_dim API instead."
+        )
+
+    def __new__(cls):
+        bare_instance = object.__new__(cls)
+        return bare_instance
+
+    @classmethod
+    def _create_constraint(
```
Comment (avikchaudhuri) [inline torch/export/__init__.py:611]: rename to just _create()

File: torch/export/__init__.py [hunk 2]
```diff
@@ -597,11 +595,34 @@ class Constraint(_ConstraintTarget):
...
     """
 
-    # NOTE(avik): In the future, this could be Union[StrictMinMaxConstraint, <other kinds>]
     constraint_range: StrictMinMaxConstraint
-    # Represent that `constraint_range` is shared with another _ConstraintTarget, which
-    # typically arises because of a specified equality with another dynamic dimension.
-    shared: Optional[_ConstraintTarget] = None
+    shared: Optional[_ConstraintTarget]
+
+    def __init__(self):
```
Comment (avikchaudhuri) [inline torch/export/__init__.py:601]: A modern way of doing this is as follows: https://replit.com/@AvikChaudhuri/MustyDistantComputer#main.py

Basically you use a metaclass and override its `__call__` method, and then you don't have to mess with the __init__ at all.

**Review by avikchaudhuri — COMMENTED** [review]
Suggesting a cleaner way of doing this.

**Pull request closed without merging**